File: src/main/java/de/fraunhofer/aisec/codyze/analysis/FindingDescription.java

package de.fraunhofer.aisec.codyze.analysis;

import com.jayway.jsonpath.Configuration;
import com.jayway.jsonpath.DocumentContext;
import com.jayway.jsonpath.JsonPath;
import com.jayway.jsonpath.Option;
import com.jayway.jsonpath.PathNotFoundException;
import org.checkerframework.checker.nullness.qual.NonNull;
import org.checkerframework.checker.nullness.qual.Nullable;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.File;
import java.io.IOException;
import java.util.List;
import java.util.Map;

/**
 * Human-readable description of a finding (i.e. of an analysis result).
 */
public class FindingDescription {
	private static final Logger log = LoggerFactory.getLogger(FindingDescription.class);

	private static FindingDescription instance = null;
	private DocumentContext json;

	private FindingDescription() {
		/* do not instantiate */
	}

	/**
	 * Singleton.
	 *
	 * Currently, there can be only one set of descriptions. This may change in the future when multiple MARK sources are supported.
	 *
	 * @return
	 */
	public static FindingDescription getInstance() {
		if (instance == null) {
			instance = new FindingDescription();
		}
		return instance;
	}

	@Nullable
	public String getDescriptionFull(@NonNull final String onFailId) {
		try {
			return json.read("$['" + onFailId + "']['fullDescription']['text']");
		}
		catch (PathNotFoundException | NullPointerException e) {
			return null;
		}
	}

	@Nullable
	public String getDescriptionShort(@NonNull final String onFailId) {
		try {
			return json.read("$['" + onFailId + "']['shortDescription']['text']");
		}
		catch (PathNotFoundException | NullPointerException e) {
			return null;
		}
	}

	@Nullable
	public String getHelpUri(@NonNull final String onFailId) {
		try {
			return json.read("$['" + onFailId + "']['helpUri']");
		}
		catch (PathNotFoundException | NullPointerException e) {
			return null;
		}
	}

	@Nullable
	public List<String> getFixes(@NonNull final String onFailId) {
		try {
			return json.read("$['" + onFailId + "']['fixes'][*]['description']['text']");
		}
		catch (PathNotFoundException | NullPointerException e) {
			return null;
		}
	}

	public void init(@NonNull File descriptionFile) {
		log.info("Parsing MARK description file from {}", descriptionFile.getAbsolutePath());
		try {
			Configuration conf = Configuration.defaultConfiguration();
			conf.addOptions(Option.DEFAULT_PATH_LEAF_TO_NULL);
			conf.addOptions(Option.SUPPRESS_EXCEPTIONS); // Return null instead of exception

			json = JsonPath.using(conf).parse(descriptionFile);

			log.info("Loaded {} descriptions", ((Map) json.read("$")).size());
		}
		catch (IOException e) {
			log.warn("Failed to load findingDescription.json, explanations will be empty", e);
		}
	}
}

File: src/main/java/de/fraunhofer/aisec/codyze/analysis/FindingDescription.kt
package de.fraunhofer.aisec.codyze.analysis

import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import com.fasterxml.jackson.module.kotlin.readValue
import java.io.File
import java.io.IOException
import org.slf4j.LoggerFactory

/** Human-readable description of a finding (i.e. of an analysis result). */
class FindingDescription private constructor() {
    private var items: Map<String, FindingDescriptionItem>? = null

    class FindingDescriptionItem(
        val helpUri: String?,
        val fullDescription: Description?,
        val shortDescription: Description?,
        val passMessage: Description?,
        val fixes: List<Fix>?
    )

    class Fix(val description: Description)
    class Description(val text: String)

    fun get(id: String): FindingDescriptionItem? {
        return items?.get(id)
    }

    fun getDescriptionFull(id: String): String? {
        return items?.get(id)?.fullDescription?.text
    }

    fun getDescriptionShort(id: String): String? {
        return items?.get(id)?.shortDescription?.text
    }

    fun getDescriptionPass(id: String): String? {
        return items?.get(id)?.passMessage?.text
    }

    fun getHelpUri(id: String): String? {
        return items?.get(id)?.helpUri
    }

    fun getFixes(id: String): List<String>? {
        return items?.get(id)?.fixes?.map { it.description.text }
    }

    fun init(descriptionFile: File) {
        log.info("Parsing MARK description file from {}", descriptionFile.absolutePath)
        try {
            val mapper = jacksonObjectMapper()
            val map = mapper.readValue<Map<String, FindingDescriptionItem>>(descriptionFile)
            log.info("Loaded {} description(s)", map.size)

            items = map
        } catch (e: IOException) {
            log.warn("Failed to load findingDescription.json, explanations will be empty", e)
        }
    }

    companion object {
        private val log = LoggerFactory.getLogger(FindingDescription::class.java)

        /**
         * Singleton.
         *
         * Currently, there can be only one set of descriptions. This may change in the future when
         * multiple MARK sources are supported.
         *
         * @return
         */
        @JvmStatic val instance: FindingDescription = FindingDescription()
    }
}
